### ISSUE TITLE
Empty cells of of an empty notebook are deleted with first save

### ISSUE THREAD

**Issue by marckruzik** [dotnet/interactive#2803]
### Describe the bug

* create an empty .dib C# Notebook
* the Notebook is created with a first empty cell
* add another empty cell by clicking on the "+ Code"
* the Notebook has 2 empty cells
* save the Notebook using CTRL+S
* the Notebook is reloaded in the same tab
* the Notebook has only 1 cell

### Screenshots

![empty_cell](https://user-images.githubusercontent.com/51711029/223506179-a32c551d-400b-4630-9280-756fe3419e30.gif)

### Please complete the following:

**Which version of .NET Interactive are you using?** (In a notebook, run the `#!about` magic command. ):
Version: 1.0.410101+17208a3471a8097c7b86e0e0be98840931cea281

- OS
    - [ ] Windows 11
    - [x] Windows 10
    - [ ] macOS
    - [ ] Linux (Please specify distro)
    - [ ] iOS
    - [ ] Android
 - Browser 
    - [ ] Chrome
    - [ ] Edge
    - [ ] Firefox
    - [ ] Safari
 - Frontend
    - [ ] Jupyter Notebook
    - [ ] Jupyter Lab
    - [ ] nteract
    - [x] Visual Studio Code 
    - [ ] Visual Studio Code Insiders
    - [ ] Visual Studio
    - [ ] Other (please specify)

**Comment by colombod** [comment]
This happens because the dub format merges consecutive blocks with the same language

**Comment by marckruzik** [comment]
Ok, I understand the intended behavior.

During my work, I saw empty cells staying in the notebook, so I was thinking the .dib notebooks were able to save empty cells.
But I guess it's some kind of cache bug.
I made some tests, and I can actually reproduce a bug.

I create a Notebook with a cell containing a simple C# code, then:
* I add 2 empty cells at the bottom
* I save the Notebook
* I close the Notebook
* I open the Notebook
* the 2 empty cells have disappear
* I do all the above again (re-adding 2 empty cells at the bottom)

Here is an example where I open a Notebook I modified following the above procedure several times. It contains empty cells. Then I add and remove empty cells and save the notebook (CTRL+S). The right part (Notepad++) displays the content of the file.
Gif:
![empty_cell_remaining](https://user-images.githubusercontent.com/51711029/223589643-fe4a485d-d272-4c40-bbe0-6305f184daf9.gif)

So here, I reached a point where:
* I can put empty cells anywhere
* the empty cells are not removed when closing and opening the file
* the content of the file is correct (code modified is changed, no empty cell, no character returns or other things)

It looks like VS Code keeps in memory the empty cells.
Closing and reopening VS Code make the empty cells go away.

**Comment by marckruzik** [comment]
I installed the last version of the polyglot notebook extension and VS Code, the problem still occurs.
(first post edited to put the last version number)

**Comment by colombod** [comment]
With dib files it is by design. What is the scenario you are trying to implement with empty cells? Ipynb would support that .

**Comment by marckruzik** [comment]
I don't try to implement something with empty cells.
That's a bug I saw, usually by leaving empty cells at the end of the file,
and seeing them either kept or removed, therefore creating an inconsistent
behavior.

On a more general note, I know minor bugs like this one are often
overlooked, especially when it's related to cache, or happening during the
first 5 seconds.
But I prefer to report them, as they can hide more complex issues.

**Comment by colombod** [comment]
But with dib is by design, it is not a bug, but worth some decent documentation to explain the file format and the implications

**Comment by marckruzik** [comment]
Yes, I understand the intended behavior is to delete empty cells. So I would like empty cells to be deleted everytime.
But as I explained earlier, empty cells stay in cache:
https://github.com/dotnet/interactive/issues/2803#issuecomment-1459098284
So I think those remaining empty cells are a bug and they should be removed.

**Comment by marckruzik** [comment]
What should I do?
Keeping this issue as "please add more documentation about dib behavior with empty cells", and create another issue for "empty cells are kept in cache instead of being removed"?

**Comment by colombod** [comment]
I am not sure I understand what you mean. Opening an issue to document file format sounds a good idea to me. the second part of your suggestion less clear. Dib will collapse empty fragment by design, I am not clear what is the scenario that needs to create a set of empty code cells all for the same kernel and persist them. What will the user do with that dib file?

**Comment by marckruzik** [comment]
> Opening an issue to document file format sounds a good idea to me. 

I agree.

> Dib will collapse empty fragment by design

I agree.

> I am not sure I understand what you mean

There is some kind of bug cache, because sometimes when I open a notebook, there are empty cells. This is a bug.

Here is a shorter version of the gif from https://github.com/dotnet/interactive/issues/2803#issuecomment-1459098284 where I open the notebook, and we can see empty cells.

![ezgif-5-2e6abb2c95](https://user-images.githubusercontent.com/51711029/224376476-6d5edb66-954e-461a-8aee-7823c8e3f552.gif)

When I open a .dib notebook, there should be no empty cells, right?

Here is a shorter description of the bug:
* The first few times I save and close a notebook with empty cells, at the next opening the empty cells disappear -> this is ok.
* Then the bug happens, and empty cells stay at each opening (even when closing and reopening the notebook) -> this is a bug.

**Comment by marckruzik** [comment]
> Have you opened , linked this in the vscode repository? we do not control that part of api.

I did not open another issue. I use VS Code only for notebooks, therefore I post my issues here.

**Comment by jonsequitur** [comment]
@marckruzik Setting aside the cache/display issue for a moment, which would you say is preferable?

* Empty cells are removed on save and reopen
* Empty cells are preserved on save and reopen
* No preference

**Comment by jonsequitur** [comment]
> Dib will collapse empty fragment by design, I am not clear what is the scenario that needs to create a set of empty code cells all for the same kernel and persist them. 

I think this is more of an emergent behavior than by design. It's the way the code happens to work, but the decision wasn't discussed heavily, and I can certainly see it as being unintuitive.

**Comment by marckruzik** [comment]
The interface allows to create empty cells, and to configure them.
So I would say:
* Empty cells are preserved on save and reopen

I am going to explain the kind of situation which confused me.

Example:
- There is a .dib notebook with 1 C# cell.
- I open the notebook.
- I put an empty cell at the end of the file, and set it as markdown, because I intend to write some explanation.
- The notebook is marked as edited.
- I save my work.
- I close the notebook
- I reopen the notebook
- My last cell (the empty markdown cell) has disappeared.
- I am confused, as I edited the notebook, see it as edited, and saved it.

Other confusion: 
* a cell is considered empty when it contains character returns -> it disappears
* a cell is not considered empty when it contains a space -> it stays

I say this, because when working with markdown (and copy-paste), it's easy to leave a character return or a space. So I was not understanding why the empty cells were disappearing sometimes (and sometimes not).

Other confusion:
* Sometimes VS Code should make empty cells disappear, but they are preserved (what I explained here https://github.com/dotnet/interactive/issues/2803#issuecomment-1464104365 )

**Conclusion**
1. From a user point of view, the most logical behavior would be for empty cells to be preserved on save and reopen. 
2. From a user point of view, deleting empty cells is closer to a "cleaning operation".

I can add that I don't know any code or word editing software which actually deletes your empty lines or empty paragraphs. So I think the polyglot extension should preserve empty cells.
